cedar-policy/src/api/id.rs

```diff
@@ -0,0 +1,289 @@
+    /// ```
+    /// # use cedar_policy::{Entity, EntityId, EntityTypeName, EntityUid};
+    /// # use std::str::FromStr;
+    /// let json_data = serde_json::json!({ "__entity": { "type": "User", "id": "alice" } });
+    /// let euid = EntityUid::from_json(json_data).unwrap();
+    /// assert_eq!(euid.type_name(), &EntityTypeName::from_str("User").unwrap());
+    /// ```
+    pub fn type_name(&self) -> &EntityTypeName {
+        // PANIC SAFETY by invariant on struct
+        #[allow(clippy::panic)]
+        match self.0.entity_type() {
+            ast::EntityType::Unspecified => panic!("Impossible to have an unspecified entity"),
+            ast::EntityType::Specified(name) => EntityTypeName::ref_cast(name),
+        }
+    }
+
+    /// Returns the id portion of the Euid
+    /// ```
+    /// # use cedar_policy::{Entity, EntityId, EntityTypeName, EntityUid};
+    /// # use std::str::FromStr;
+    /// let json_data = serde_json::json!({ "__entity": { "type": "User", "id": "alice" } });
+    /// let euid = EntityUid::from_json(json_data).unwrap();
+    /// assert_eq!(euid.id(), &EntityId::from_str("alice").unwrap());
+    /// ```
+    pub fn id(&self) -> &EntityId {
+        EntityId::ref_cast(self.0.eid())
+    }
+
+    /// Creates `EntityUid` from `EntityTypeName` and `EntityId`
+    ///```
+    /// # use cedar_policy::{Entity, EntityId, EntityTypeName, EntityUid};
+    /// # use std::str::FromStr;
+    /// let eid = EntityId::from_str("alice").unwrap();
+    /// let type_name: EntityTypeName = EntityTypeName::from_str("User").unwrap();
+    /// let euid = EntityUid::from_type_name_and_id(type_name, eid);
+    /// # assert_eq!(euid.type_name(), &EntityTypeName::from_str("User").unwrap());
+    /// # assert_eq!(euid.id(), &EntityId::from_str("alice").unwrap());
+    /// ```
+    pub fn from_type_name_and_id(name: EntityTypeName, id: EntityId) -> Self {
+        // INVARIANT: `from_components` always constructs a Concrete id
+        Self(ast::EntityUID::from_components(name.0, id.0, None))
+    }
+
+    /// Testing utility for creating `EntityUids` a bit easier
+    #[cfg(test)]
+    pub(crate) fn from_strs(typename: &str, id: &str) -> Self {
+        Self::from_type_name_and_id(
+            EntityTypeName::from_str(typename).unwrap(),
+            EntityId::from_str(id).unwrap(),
+        )
+    }
+
+    /// Construct an [`EntityUid`] from the internal type.
+    /// This function is only intended to be used internally.
+    pub(crate) fn new(uid: ast::EntityUID) -> Self {
+        Self(uid)
+    }
+
+    /// Deconstruct an [`EntityUid`] to get the internal type.
+    /// This function is only intended to be used internally.
+    pub(crate) fn into_inner(self) -> ast::EntityUID {
+        self.0
+    }
+
+    /// Deconstruct an [`EntityUid`] to get the internal type.
+    /// This function is only intended to be used internally.
+    pub(crate) fn as_inner(&self) -> &ast::EntityUID {
+        &self.0
+    }
```

Comment:
Thoughts on making this an `impl AsRef` instead of (or in addition to) `as_inner`?

I guess the same goes for `into_inner()` it could be an `impl From` instead / in addition

Reply:
Rust noob question: Can I make the `AsRef` and `From` impls `pub(crate)`? I don't want them exposed through `cedar_policy`'s main interface since they expose internal types.

Reply:
I don't believe you can.  We put up with this for `From` impls on error types, but mark them `doc(hidden)`.  We could do the same here.  But I am also perfectly fine with keeping this as `into_inner()` and `as_inner()`, was just asking for thoughts 😄